File: 075/main.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function (nums) {
  const len = nums.length
  let p1 = 0
  let p2 = len - 1
  let i = 0
  while (i <= p2) {
    if (nums[i] === 0) {
      nums[i] = nums[p1]
      nums[p1++] = 0
    } else if (nums[i] === 2) {
      swap(i, p2)
    }
    i++
  }
  return nums

  function swap (a, b) {
    let tmp = nums[a]
    nums[a] = nums[b]
    nums[b] = tmp
    p2 = b - 1
    if (nums[a] === 2 && a < p2 + 1) {
      swap(a, p2)
    } else if (nums[a] === 0 && a >= p1) {
      nums[a] = nums[p1]
      nums[p1++] = 0
    }
  }
}

// console.log(sortColors([1, 0, 0, 2, 1, 0, 1, 0, 0, 1, 2, 2, 1, 2, 0, 1]))
// console.log(sortColors([2, 2]))
console.log(sortColors([2, 2, 0, 0, 2, 0, 2, 1, 0]))

File: 075/main.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    void sortColors(vector<int>& nums) {
        int p1 = 0;
        int p2 = nums.size() - 1;
        while (nums[p1] == 0) p1++;
        while (nums[p2] == 2) p2--;
        int i = p1;
        while (i <= p2 && p1 <= p2) {
            if (nums[i] == 0) {
                nums[i++] = nums[p1];
                nums[p1] = 0;
                while (nums[p1] == 0) p1++;
            } else if (nums[i] == 2) {
                nums[i] = nums[p2];
                nums[p2] = 2;
                while (nums[p2] == 2) p2--;
            } else i++;
        }
    }
};

int main()
{
    Solution sol;
    vector<int> nums;
    nums.push_back(2);
    nums.push_back(0);
    nums.push_back(0);
    sol.sortColors(nums);
    return 0;
}
